File: service/benchmarks/munchausen.lua
local skynet = require("skynet")
local amount = ...
amount = tonumber(amount)

skynet.start(function()

    print("======= begin test =======")
    
    local find_amount = 0
    local function isMunchausen(n)
        local sum, digit, acc = 0, 0, n
        while acc > 0 do
            digit = acc % 10.0
            sum = sum + digit ^ digit
            acc = acc // 10 -- integer div
        end
        return sum == n
    end

    local ts_begin = skynet.now() * 10;
    for i = 1, amount do
        if isMunchausen(i) then
            find_amount = find_amount + 1
        end
    end
    local ts_end = skynet.now() * 10;
    
    print(string.format("\n\nlua munchausen result:%d \tamount: %d\t\tms:%d\n\n", find_amount, amount, ts_end - ts_begin))
    print("======= end test =======")

end)

File: demo/service/benchmarks/munchausen.ts
import * as skynet from "skynet"

skynet.start(() => {
    let [service_name, amount_str] = JS_INIT_ARGS.split(" ");
    let amount = Number(amount_str);

    console.log("======= begin test =======")

    let find_amount = 0;
    function isMunchausen1(n: number) {
        let sum = 0;
        let digit = 0;
        let acc = n;
        while (acc > 0) {
            digit = acc % 10
            sum = sum + Math.pow(digit, digit)
            acc = Math.floor(acc / 10)
        }
        return sum == n
    }

    let isMunchausen2 = (n: number) =>
        n.toString()
            .split('')
            .reduce(
                (a, c) => (
                    d => a + Math.pow(d, d)
                )(parseInt(c, 10)),
                0
            ) === n;

    let ts_begin = Date.now();
    for (let i = 1; i <= amount; i++) {
        if (isMunchausen1(i)) {
            find_amount++;
        }
    }
    let ts_end = Date.now();

    console.log(`\n\nv8 munchausen result:${find_amount} \tamount:${amount}\t\tms:${ts_end - ts_begin}\n\n`);
    console.log("======= end test =======")
})